src/index.tsx

```diff
@@ -306,3 +343,110 @@ const getSystemTheme = (e?: MediaQueryList | MediaQueryListEvent) => {
+try {
+  const cookies = require('next/headers').cookies();
+  getCookie = (name: string) => cookies.get(name)?.value;
+} catch(e) {}
+
+// Properties for rendering <html> on the server in a way that will match client after hydration
+const getThemeHtmlProps = ({
+  attribute = 'data-theme',
+  cookieName = 'theme',
+  defaultTheme = 'light',
+  enableColorScheme = true,
+  value,
+}: ThemeProviderProps) => {
+  const props: HTMLProps<HTMLHtmlElement> = {}
+
+  const resolved = getCookie?.(cookieName) || defaultTheme
+  const name = value ? value[resolved] : resolved
+
+  if (attribute === 'class') {
+    if (name) props.className = name
+  } else {
+    (props as Record<string, string>)[attribute] = name
+  }
+
+  if (enableColorScheme) {
+    const fallback = colorSchemes.includes(defaultTheme) ? defaultTheme : null
+    const colorScheme = colorSchemes.includes(resolved) ? resolved : fallback
+    props.style = { colorScheme } as CSSProperties
+  }
+
+  return props
+}
+
+// Wraps an <html> element on the server to apply themes before reaching the client
+export const ServerThemeProvider: React.FC<ThemeProviderProps> = ({ children, ...props }) => {
+  if (!children || (children as ReactElement).type !== 'html') {
+    throw new Error('<ServerThemeProvider> must contain the <html> element.')
+  }
+  const child = children as ReactElement
+  const original = child.props
+  const resolved = getThemeHtmlProps(props)
+  if (original.className && resolved.className) {
+    resolved.className = `${original.className} ${resolved.className}`
+  }
+  if (original.style && resolved.style) {
+    let originalStyle = original.style
+    if (typeof originalStyle === 'string') {
+      originalStyle = Object.fromEntries(originalStyle.split(/;\s*/).map(x => x.split(/\s*:\s*/)))
+    }
+    resolved.style = {
+      ...resolved.style,
+      ...originalStyle,
+    }
+  }
+
+  let newKids: ReactNodeArray
+  if (!original.children) {
+    newKids = []
+  } else if (!Array.isArray(original.children)) {
+    newKids = [original.children]
+  } else {
+    newKids = [...original.children]
+  }
+  let bodyIndex = newKids.findIndex(x => x && typeof x === 'object' && (x as ReactElement).type === 'body')
+  let body: ReactElement
+  if (bodyIndex !== -1) {
+    body = newKids.splice(bodyIndex, 1)[0] as ReactElement
+  } else {
+    bodyIndex = newKids.length
+    body = <body />
+  }
+  let bodyChildren: ReactNodeArray
+  if (!body.props?.children) {
+    bodyChildren = []
+  } else if (!Array.isArray(body.props.children)) {
+    bodyChildren = [body.props.children]
+  } else {
+    bodyChildren = [...body.props.children]
+  }
+  const attrs = !props.value ? (props.themes || defaultThemes) : Object.values(props.value)
+  const defaultTheme = props.enableSystem !== false ? 'system' : 'light'
+  resolved.children = [
+    ...newKids.slice(0, bodyIndex),
+    cloneElement(body, {
+      children: [
+        ...bodyChildren,
```

Comment:
This should probably be spread at the end instead of the beginning here to match what `next-themes` is currently doing: Blocking.

If not you will see a short flash of white/dark depending on the default value.

Without this PR via wrapping the layout children in a "use client" Provider component: <https://discord-e36c7a2o4-discordjs.vercel.app/>

With this PR wrapping the HTML element with a ServerThemeProvider: <https://discord-lc209wwm5-discordjs.vercel.app/>